daemon/src/maker_cfd.rs

```diff
@@ -395,6 +334,94 @@ where
     }
 }
 
+#[xtra_productivity]
+impl<O, M, T, W> Actor<O, M, T, W> {
+    async fn handle_accept_rollover(&mut self, msg: AcceptRollOver) -> Result<()> {
+        if !self
+            .rollover_actors
+            .send(&msg.order_id, rollover_maker::AcceptRollOver)
+            .await
+        {
+            tracing::warn!(%msg.order_id, "No active rollover");
+        }
+
+        Ok(())
+    }
+}
+
+impl<O, M, T, W> Actor<O, M, T, W>
+where
+    O: xtra::Handler<oracle::GetAnnouncement> + xtra::Handler<oracle::MonitorAttestation>,
+    M: xtra::Handler<monitor::StartMonitoring>,
+    T: xtra::Handler<maker_inc_connections::TakerMessage>
+        + xtra::Handler<Stopping<rollover_maker::Actor>>,
+    W: 'static,
+    Self: xtra::Handler<Stopping<rollover_maker::Actor>>,
+{
+    async fn handle_propose_roll_over(
+        &mut self,
+        proposal: RollOverProposal,
+        taker_id: Identity,
+        ctx: &mut Context<Self>,
+    ) -> Result<()> {
+        tracing::info!(
+            "Received proposal from the taker {}: {:?} to roll over order {}",
+            taker_id,
+            proposal,
+            proposal.order_id
+        );
+
+        // check if CFD is in open state, otherwise we should not proceed
+        let mut conn = self.db.acquire().await?;
+        let cfd = load_cfd_by_order_id(proposal.order_id, &mut conn).await?;
+        match cfd {
+            Cfd {
+                state: CfdState::Open { .. },
+                ..
+            } => (),
+            _ => {
+                anyhow::bail!("Order is in invalid state. Cannot propose roll over.")
+            }
+        };
+
+        let this = ctx.address().expect("acquired own address");
+
+        let (rollover_actor_addr, _rollover_actor_future) = rollover_maker::Actor::new(
+            &self.takers,
+            cfd,
+            taker_id,
+            self.oracle_pk,
+            &this,
+            &self.oracle_actor,
+            (&self.takers, &this),
+            self.projection_actor.clone(),
+        )
+        .create(None)
+        .run();
+
+        self.rollover_actors
+            .insert(proposal.order_id, rollover_actor_addr);
+
+        let new_proposal = UpdateCfdProposal::RollOverProposal {
+            proposal: proposal.clone(),
+            direction: SettlementKind::Incoming,
+        };
+
+        self.projection_actor
+            .send(try_into_update_rollover_proposal(new_proposal)?)
+            .await?;
```

Comment:
Can we move this into the `started` hook of the actor?

Reply:
to do this we will have pass the proposal into the actor. Not sure if that is a good idea. I am seeing the proposal and the rollover as different domain objects.

Reply:
The proposal is part of the rollover protocol. In an ideal world, I would like to kick off the protocol directly from the user interaction and perhaps even move the loading of the CFD into the database into that actor as well. The entire process should be encapsulated in an actor.